File: Homework/Ngan_Luu/845-longest-mountain-in-array.py
# https://leetcode.com/problems/longest-mountain-in-array/
class Solution:
    def longestMountain(self, arr: List[int]) -> int:
        slow=0
        res=0
        while slow<len(arr):
            fast=slow
            if fast+1<len(arr) and arr[fast]<arr[fast+1]:
                while fast+1<len(arr) and arr[fast]<arr[fast+1] :
                    fast+=1
                if  fast+1<len(arr) and arr[fast]>arr[fast+1]:
                    while fast+1<len(arr) and arr[fast]>arr[fast+1]:
                        fast+=1
                    res=max(res,fast-slow+1)
            slow=max(fast,slow+1)
        return res

File: Homework/Hoang_Nguyen/lesson11/845_longestMountainInArray.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var longestMountain = function (arr) {
  const n = arr.length;
  let res = 0;
  let i = 0;
  while (i < n) {
    let top = i;
    while (top + 1 < n && arr[top + 1] > arr[top]) {
      top++;
    }
    let bottom = top;
    while (bottom + 1 < n && arr[bottom + 1] < arr[bottom]) {
      bottom++;
    }
    if (i < top && top < bottom) {
      res = Math.max(res, bottom - i + 1);
    }

    if (i < bottom) {
      i = bottom;
    } else {
      i++;
    }
  }
  return res;
};
